### ISSUE TITLE
stb_textedit.h crash bug in stb_textedit_discard_redo()

### ISSUE THREAD

**Issue by ocornut** [nothings/stb#321]
As pointed on twitter, just posting here for reminder/reference.

In stb_textedit_discard_redo()

```
-         STB_TEXTEDIT_memmove(state->undo_char + state->redo_char_point, state->undo_char + state->redo_char_point-n, (size_t) ((size_t)(STB_TEXTEDIT_UNDOSTATECOUNT - state->redo_char_point)*sizeof(STB_TEXTEDIT_CHARTYPE)));
+         STB_TEXTEDIT_memmove(state->undo_char + state->redo_char_point, state->undo_char + state->redo_char_point-n, (size_t) ((size_t)(STB_TEXTEDIT_UNDOCHARCOUNT - state->redo_char_point)*sizeof(STB_TEXTEDIT_CHARTYPE)));
```

This would haven't triggered so much as stb_textedit_discard_redo() is rather infrequently called (it does get called when we consume more than STB_TEXTEDIT_UNDOCHARCOUNT characters worth of undo records.

**Comment by chrisws** [comment]
Hi,

I found the same issue and this was my fix:

         short count = state->redo_char_point > STB_TEXTEDIT_UNDOSTATECOUNT ? STB_TEXTEDIT_UNDOSTATECOUNT : state->redo_char_point;
         STB_TEXTEDIT_memmove(state->undo_char + state->redo_char_point, state->undo_char + state->redo_char_point-n, (size_t) ((STB_TEXTEDIT_UNDOSTATECOUNT - count)*sizeof(STB_TEXTEDIT_CHARTYPE)));

This avoid passing a negative size value to memmove. Not sure whether this is the full correct fix or just masking some other issue.

**Comment by ocornut** [comment]
TL;DR; you might already have the same fix locally in your version at work. Might be easier to just check for that before engaging with that bug.

@chrisws That will prevent your software from crashing but still incorrect.

@nothings Following the twitter conversation 3 weeks ago I had another bug reported
https://github.com/ocornut/imgui/issues/715

Basically the test sequence that is natural to repro within the ImGui demo is:
- STB_TEXTEDIT_UNDOCHARCOUNT = 999
- initial buffer is 350 long
- copy and paste 3 times, buffer is now 350*4 = 1400 char long, not using the char buffer yet.
- undo 2 times, ok, 350x2 char used in the char buffer
- undo a 3rd time: we don't have enough room in char buffer, `stb_textedit_discard_redo` is called. 

This is merely an example and of course it is just a case of discard_redo being broken.

I'm quite confused by the undo-stack code but I _think_ I came up with two fixes.

A. In `stb_text_redo`
```
   if (r.insert_length) {
      // easy case: need to insert n characters
      STB_TEXTEDIT_INSERTCHARS(str, r.where, &s->undo_char[r.char_storage], r.insert_length);
++      s->redo_char_point += r.insert_length;
   }
```

Without this fix, a simple sequence of Paste, Undo/Redo/Undo/Redo/... would exhaust the character buffer.

B. in `stb_textedit_discard_redo`
Replace
```
++state->redo_point;
STB_TEXTEDIT_memmove(state->undo_rec + state->redo_point-1, state->undo_rec + state->redo_point, (size_t) ((size_t)(STB_TEXTEDIT_UNDOSTATECOUNT - state->redo_point)*sizeof(state->undo_rec[0])));
```
By
```
STB_TEXTEDIT_memmove(state->undo_rec + state->redo_point, state->undo_rec + state->redo_point-1, (size_t) ((size_t)(STB_TEXTEDIT_UNDOSTATECOUNT - state->redo_point)*sizeof(state->undo_rec[0])));
++state->redo_point;
```

This is coming from tracing into the code and displaying the state on screen. This is the sequence being executed BEFORE the fix:

![textedit_before](https://cloud.githubusercontent.com/assets/8225057/16379366/8f379d6a-3c71-11e6-86ae-3e0ab6e62f8c.gif)

Select all, copy, unselect, paste 3 times, undo 3 times, redo
The third undo calls `stb_textedit_discard_redo` and you can see the entries at the bottom looking invalid.

The final redo triggers character insertion at an invalid position. Note that the assert at the end is the code in my `STB_TEXTEDIT_INSERTCHARS` implementation asserting that `pos <= STB_TEXTEDIT_STRINGLEN(obj)`. Some other implementation of the INSERTCHARS function may silently return and then the bug wouldn't be so noticeable (but undo/redo stack still appears broken).

And AFTER the fix:

![textedit_after](https://cloud.githubusercontent.com/assets/8225057/16379371/92646c2a-3c71-11e6-8d42-d01ec8fd445d.gif)

Basically as I understand stb_textedit_discard_redo was just very wrong. It's very easy to overlook because STB_TEXTEDIT_UNDOCHARCOUNT can increased, and it requires exhausting the buffer and using undo+redo.

**Comment by chrisws** [comment]
I finally got around to testing the suggested fixes - Fix A (In stb_text_redo) works for me and seems most logically correct.

**Comment by ocornut** [comment]
My wording was perhaps ambiguous, but I meant that both fixes are desirable and required.

**Comment by chrisws** [comment]
Hi,

Thanks so much for your kind update. 

I *think* there may be one more bug in stb_textedit_discard_redo() - in this line:
         STB_TEXTEDIT_memmove(state->undo_char + state->redo_char_point, state->undo_char + state->redo_char_point-n, (size_t) ((STB_TEXTEDIT_UNDOSTATECOUNT - state->redo_char_point)*sizeof(STB_TEXTEDIT_CHARTYPE)));

shouldn't STB_TEXTEDIT_UNDOSTATECOUNT instead be STB_TEXTEDIT_UNDOCHARCOUNT ? 

This is with both of your changes applied to version 1.11. On my system it makes the difference between crashing or not.

**Comment by ocornut** [comment]
Correct. This is actually what the first post of this thread discusses.

For what it is worth, my modified version is located here:
https://github.com/ocornut/imgui/blob/master/stb_textedit.h

**Comment by nothings** [comment]
I believe the correct fix for B is

      STB_TEXTEDIT_memmove(state->undo_rec + state->redo_point+1, state->undo_rec + state->redo_point, (size_t) ((STB_TEXTEDIT_UNDOSTATECOUNT - state->redo_point)*sizeof(state->undo_rec[0])));

The range of redo records is [redo_point, STATECOUNT). After deleting the last one, we want to move all the remaining ones down towards the end (because it's a stack growing downwards). So @ocornut's fix moves one extra item (the item at redo_point-1, to redo_point). Most of the time this is harmless because the slot at redo_point WAS just being used by redo, but is about to be unused, so moving something into it should have no effect.

However, if you ever got redo_point to 0 (which I *think* is possible), it would be moving from outside the array, undo_rec[-1]. This will access earlier fields in the TexteditState struct, so *should* be harmless, but might as well fix it right.

**Comment by nothings** [comment]
Fixed (untested!) in next version.

**Comment by ocornut** [comment]
FYI Sean (as you are on a stb_ streak right now :)
I realize this thread is overally quite confusing with no clear call-to-action. 

I worked on this two days ago and thought I fixed the last issue and was about to submit a "definitive report" but a user is still reporting an issue with static analyzer. Will report back when I know more. You may want to skip this right now if it is too confusing.

**Comment by nothings** [comment]
No worries, as I'm only doing PRs not issues as that was enough of a workload :(

**Comment by chrisws** [comment]
If I can help in anyway let me know (testing etc)

**Comment by rygorous** [comment]
The crash appears to be fixed by Omar's PR and the discussion thread for the PR notes that static analysis appears to be overly picky here, so I don't think there's anything left to do on this. Closing.

**Comment by ocornut** [comment]
FYI (may or not be related to this)
There was another bug report on our end which I haven't investigated yet:
https://github.com/ocornut/imgui/issues/4038
As our version has various mods it may be our issue or may be a stock stb_textedit issue.
I'll try to investigate the linked issue and report here (this or new issue depending on result).
